diff --git a/src/compiler/crystal/program.cr b/src/compiler/crystal/program.cr
index e6f9dedbb..278dfafcf 100644
--- a/src/compiler/crystal/program.cr
+++ b/src/compiler/crystal/program.cr
@@ -81,10 +81,10 @@ module Crystal
 
       @types["Macro"] = macro_mod = NonGenericModuleType.new self, self, "Macro"
       macro_var = macro_mod.types["Var"] = NonGenericClassType.new self, macro_mod, "Var", @reference
-      macro_var.lookup_instance_var("@padding").type = PaddingType.new(self, 24)
+      macro_var.lookup_instance_var("@padding").type = PaddingType.new(self, 17)
 
       macro_ivar = macro_mod.types["InstanceVar"] = NonGenericClassType.new self, macro_mod, "InstanceVar", @reference
-      macro_ivar.lookup_instance_var("@padding").type = PaddingType.new(self, 24)
+      macro_ivar.lookup_instance_var("@padding").type = PaddingType.new(self, 17)
 
       @symbols = Set(String).new
       @global_vars = {} of String => Var
